File: client-android/lib/src/main/kotlin/stasis/client_android/lib/discovery/ServiceDiscoveryResult.kt
package stasis.client_android.lib.discovery

sealed class ServiceDiscoveryResult {
    abstract val asString: String

    data object KeepExisting : ServiceDiscoveryResult() {
        override val asString: String = "result=keep-existing"
    }

    data class SwitchTo(
        val endpoints: Endpoints,
        val recreateExisting: Boolean
    ) : ServiceDiscoveryResult() {
        override val asString: String by lazy {
            "result=switch-to,endpoints=${endpoints.asString},recreate-existing=$recreateExisting"
        }
    }

    data class Endpoints(
        val api: ServiceApiEndpoint.Api,
        val core: ServiceApiEndpoint.Core,
        val discovery: ServiceApiEndpoint.Discovery
    ) {
        val asString: String by lazy {
            listOf(api.id, core.id, discovery.id).joinToString(separator = ";")
        }
    }
}

File: core/src/main/scala/stasis/core/discovery/ServiceDiscoveryResult.scala
package stasis.core.discovery

sealed trait ServiceDiscoveryResult {
  def asString: String
}

object ServiceDiscoveryResult {
  case object KeepExisting extends ServiceDiscoveryResult {
    override lazy val asString: String = "result=keep-existing"
  }

  final case class SwitchTo(
    endpoints: Endpoints,
    recreateExisting: Boolean
  ) extends ServiceDiscoveryResult {
    override lazy val asString: String =
      s"result=switch-to," +
        s"endpoints=${endpoints.asString}," +
        s"recreate-existing=${recreateExisting.toString}"
  }

  final case class Endpoints(
    api: ServiceApiEndpoint.Api,
    core: ServiceApiEndpoint.Core,
    discovery: ServiceApiEndpoint.Discovery
  ) {
    lazy val asString: String = Seq(api.id, core.id, discovery.id).mkString(";")
  }
}
